File: backend/routes/files.js
const path = require('path');
const fs = require('fs');
const express = require('express');
const multer = require('multer');
const { processExcel } = require('../services/excelService');
const { generatePdf } = require('../services/pdfService');

const router = express.Router();

const rootDir = path.join(__dirname, '..');
const uploadsDir = path.join(rootDir, 'uploads');
const outputDir = path.join(rootDir, 'output');
if (!fs.existsSync(uploadsDir)) fs.mkdirSync(uploadsDir, { recursive: true });
if (!fs.existsSync(outputDir)) fs.mkdirSync(outputDir, { recursive: true });

const storage = multer.diskStorage({
  destination: (req, file, cb) => cb(null, uploadsDir),
  filename: (req, file, cb) => cb(null, Date.now() + path.extname(file.originalname)),
});
const upload = multer({ storage });

router.post('/generate-quote', upload.single('file'), async (req, res) => {
  try {
    if (!req.file) {
      return res.status(400).json({ success: false, message: 'No se ha subido ningún archivo' });
    }
    const cotizacion = await processExcel(req.file.path);
    res.json({ success: true, cotizacion });
  } catch (error) {
    console.error('Error al procesar el archivo Excel:', error);
    res.status(500).json({ success: false, message: 'Error al procesar el archivo Excel' });
  }
});

router.post('/generate-pdf', async (req, res) => {
  try {
    const { procedimiento, materiales, manoDeObra, totales } = req.body;
    const { pdfPath, filename } = await generatePdf({ procedimiento, materiales, manoDeObra, totales }, outputDir);
    res.json({ success: true, pdfPath: `/api/download-pdf/${filename}` });
  } catch (error) {
    console.error('Error en la generación del PDF:', error);
    res.status(500).json({ success: false, message: 'Error en la generación del PDF' });
  }
});

router.get('/download-pdf/:filename', (req, res) => {
  try {
    const { filename } = req.params;
    const filePath = path.join(outputDir, filename);
    if (!fs.existsSync(filePath)) {
      return res.status(404).json({ success: false, message: 'Archivo no encontrado' });
    }
    res.download(filePath, filename, (err) => {
      if (err) {
        console.error('Error al descargar el PDF:', err);
        return res.status(500).json({ success: false, message: 'Error al descargar el PDF' });
      }
    });
  } catch (error) {
    console.error('Error en la descarga del PDF:', error);
    res.status(500).json({ success: false, message: 'Error en la descarga del PDF' });
  }
});

module.exports = router;

File: Backend Python/routers/files.py
import os
import shutil
from fastapi import APIRouter, UploadFile, File, HTTPException
from fastapi.responses import FileResponse
from services.excel_service import process_excel
from services.pdf_service import generate_pdf

router = APIRouter(tags=["Files"])

UPLOAD_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', 'uploads')
OUTPUT_DIR = os.path.join(os.path.dirname(os.path.abspath(__file__)), '..', 'output')

os.makedirs(UPLOAD_DIR, exist_ok=True)
os.makedirs(OUTPUT_DIR, exist_ok=True)


@router.post("/generate-quote")
async def generate_quote(file: UploadFile = File(...)):
    try:
        file_location = os.path.join(UPLOAD_DIR, file.filename)
        with open(file_location, "wb") as buffer:
            shutil.copyfileobj(file.file, buffer)
        cotizacion = process_excel(file_location)
        return {"success": True, "cotizacion": cotizacion}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.post("/generate-pdf")
async def generate_pdf_endpoint(payload: dict):
    try:
        procedimiento = payload.get("procedimiento", {})
        materiales = payload.get("materiales", [])
        mano_de_obra = payload.get("manoDeObra", [])
        totales = payload.get("totales", {})

        pdf_path = generate_pdf(procedimiento, materiales, mano_de_obra, totales)
        filename = os.path.basename(pdf_path)
        return {"success": True, "filename": filename, "pdfPath": pdf_path}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/download-pdf/{filename}")
async def download_pdf(filename: str):
    file_path = os.path.join(OUTPUT_DIR, filename)
    if not os.path.exists(file_path):
        raise HTTPException(status_code=404, detail="PDF no encontrado")
    return FileResponse(path=file_path, filename=filename, media_type='application/pdf')